packages/byods/src/data-source-service/index.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable no-useless-catch */
+/* eslint-disable no-console */
+/* eslint-disable class-methods-use-this */
+/* eslint-disable import/no-extraneous-dependencies */
+
+import {v4 as uuidv4} from 'uuid';
+import {decodeJwt, JWTPayload} from 'jose';
+import DataSourceClient from '../data-source-client/index';
+
+export default class DataSourceService {
+  private dataSourceClient: DataSourceClient;
+  private timer: any;
+
+  /**
+   * Creates an instance of DataSourceService.
+   * @param {DataSourceClient} dataSourceClient - The  dataSourceClient instance to use for DataSourceService.
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient);
+   */
+
+  constructor(dataSourceClient: DataSourceClient) {
+    this.dataSourceClient = dataSourceClient;
+  }
+
+  /**
+   * This method refreshes the DataSource token using dataSourceId and tokenLifetimeMinute
+   * @param {string} dataSourceId The id of data source
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {Promise<void>}
+   */
+
+  public async refreshDataSourceToken(
+    dataSourceId: string,
+    tokenLifetimeMinutes?: number
+  ): Promise<void> {
```

Comment:
This should return a cancellable promise. Otherwise, we have no way to cancel the refreshing logic.

Reply:
Developer has a way to cancel the refreshing logic by creating a dataSourceService class object and invoking the stopAutoRefresh() method. This method will cancel the timer.

Reply:
The instance who has invoked the refreshDataSourceToken method will be responsible for cancelling the refreshing logic and that instance should be able to call the stopAutoRefresh() function.

Reply:
**_Addressed!_**